### ISSUE TITLE
Run arangod with "--database.auto-upgrade" option always crash silently without error log.

### ISSUE THREAD

**Issue by garyhai** [arangodb/arangodb#4160]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.3

Mode:
- [x] Single-Server

Storage-Engine:
- [x] mmfiles
- [ ] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [x] MacOS, version: 10.13.2 (17C88)

### this is an AQL-related issue:
[ ] I'm using graph features

I'm issuing AQL via:
- [ ] web interface with this browser:     running on this OS:
- [ ] arangosh
- [ ] this Driver:

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:

The issue can be reproduced using this dataset:

Please provide a way to create the dataset to run the above query on; either by a gist with an arangodump, or `db.collection.save({my: "values"}) statements. If it can be reproduced with one of the ArangoDB example datasets, it's a plus.

### this is a web interface-related issue:
I'm using the web interface with this browser:     running on this OS:
- [ ] authentication is enabled?
- [ ] using the cluster?
- [ ] _system database?

These are the steps to reproduce:
1) open the browser on http://127.0.0.1:8529
2) log in as ...
3) use database [ ] `_system` [ ] other: 
4) click to ...
...

The following problem occurs: [Screenshot?] 
Instead I would be expecting: 

### this is an installation-related issue:

Run arangod with "--database.auto-upgrade" option always crash silently without error log.
```
$> arangod --database.auto-upgrade
```

**Comment by sleto-it** [comment]
Hi @garyhai,

Thank you for opening this ticket. Please could you use the option `--log.level DEBUG` and then analyze the provided output and search for the word "upgrade"? You should find info on whether the database is already up-to-date, or not. If it is, please try to start again the server without the `--database.auto-upgrade` option, and this time it should run and stay up

I confirm however that running `arangod` with the option `--database.auto-upgrade true` exits silently and I'd said this is either a bug or something non intuitive that should be improved
 
**Result of my tests:**

1. Start 3.2.10, e.g.

```
"D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\usr\bin\arangod.exe"
```

2. Wait 3.2.10 is up and running and then stop it

3. Start 3.3.0 passing to it the 3.2.10 data directory. This is the printed output:

```
D:\>"D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\usr\bin\arangod.exe" --database.directory  D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3
2017-12-24T12:18:47Z [19632] INFO ArangoDB 3.3.0 enterprise [win64] 64bit, using VPack 0.1.30, RocksDB 5.6.0, ICU 58.1, V8 5.7.492.77, OpenSSL 1.0.2a 19 Mar 2015
2017-12-24T12:18:47Z [19632] INFO using storage engine mmfiles
2017-12-24T12:18:47Z [19632] INFO {cluster} Starting up with role SINGLE
2017-12-24T12:18:47Z [19632] INFO {authentication} Authentication is turned on (system only)
2017-12-24T12:18:47Z [19632] INFO created base application directory 'D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\var\lib\arangodb3-apps\_db'
2017-12-24T12:18:47Z [19632] INFO running WAL recovery (2 logfiles)
2017-12-24T12:18:47Z [19632] INFO replaying WAL logfile 'D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3\journals\logfile-2.db' (1 of 2)
2017-12-24T12:18:47Z [19632] INFO replaying WAL logfile 'D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3\journals\logfile-55.db' (2 of 2)
2017-12-24T12:18:47Z [19632] INFO WAL recovery finished successfully
2017-12-24T12:18:47Z [19632] ERROR In database "_system": Database directory version (30210) is lower than current version (30300).
2017-12-24T12:18:47Z [19632] ERROR In database "_system": ----------------------------------------------------------------------
2017-12-24T12:18:47Z [19632] ERROR In database "_system": It seems like you have upgraded the ArangoDB binary.
2017-12-24T12:18:47Z [19632] ERROR In database "_system": If this is what you wanted to do, please restart with the
2017-12-24T12:18:47Z [19632] ERROR In database "_system":   --database.auto-upgrade true
2017-12-24T12:18:47Z [19632] ERROR In database "_system": option to upgrade the data in the database directory.
2017-12-24T12:18:47Z [19632] ERROR In database "_system": Normally you can use the control script to upgrade your database
2017-12-24T12:18:47Z [19632] ERROR In database "_system":   /etc/init.d/arangodb stop
2017-12-24T12:18:47Z [19632] ERROR In database "_system":   /etc/init.d/arangodb upgrade
2017-12-24T12:18:47Z [19632] ERROR In database "_system":   /etc/init.d/arangodb start
2017-12-24T12:18:47Z [19632] ERROR In database "_system": ----------------------------------------------------------------------
2017-12-24T12:18:47Z [19632] FATAL Database '_system' needs upgrade. Please start the server with the --database.auto-upgrade option
2017-12-24T12:18:47Z [19632] FATAL Database '_system' upgrade failed. Please inspect the logs from the upgrade procedure
```

4. Start again 3.3.0, using the option `--database.auto-upgrade true`:

```
D:\>"D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\usr\bin\arangod.exe" --database.directory  D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3 --database.auto-upgrade true
2017-12-24T12:20:12Z [4016] INFO using storage engine mmfiles
2017-12-24T12:20:12Z [4016] INFO {authentication} Authentication is turned on (system only)
2017-12-24T12:20:13Z [4016] INFO database upgrade passed
2017-12-24T12:20:13Z [4016] INFO using endpoint 'http+tcp://127.0.0.1:8529' for non-encrypted requests
2017-12-24T12:20:15Z [4016] INFO ArangoDB (version 3.3.0 enterprise [win64]) is ready for business. Have fun!
```

**--> Reproduced that no info on the database upgrade are shown, and the process ends**

Note:

a) If I start again the process, without `--database.auto-upgrade true`, it starts normally:

```
"D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\usr\bin\arangod.exe" --database.directory  D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3 
```

b) This is what is printed with DEBUG log level (using again the `--database.auto-upgrade` option) - on a database that was upgraded already:

```
"D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\usr\bin\arangod.exe" --database.directory  D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3 --database.auto-upgrade true --log.level DEBUG

[...]
2017-12-24T12:24:35Z [15928] DEBUG running database init/upgrade
2017-12-24T12:24:35Z [15928] DEBUG In database "_system": Database is up-to-date (30300/standalone/existing)
2017-12-24T12:24:35Z [15928] DEBUG database '_system' init/upgrade done
2017-12-24T12:24:35Z [15928] INFO database upgrade passed
[...]
```

- This is what is printed with DEBUG log level (using again the `--database.auto-upgrade` option) - on a database that has still be be upgraded:

```
"D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\usr\bin\arangod.exe" --database.directory  D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3 --database.auto-upgrade true --log.level DEBUG

[...]
2017-12-24T12:28:19Z [18648] DEBUG running database init/upgrade
2017-12-24T12:28:19Z [18648] DEBUG In database "_system": Found 17 defined task(s), 4 task(s) to run
2017-12-24T12:28:19Z [18648] DEBUG In database "_system": state standalone/upgrade, tasks updateUserModels, setupApps, createAppsIndex, setupAppBundles
2017-12-24T12:28:19Z [18648] DEBUG In database "_system": upgrade successfully finished
2017-12-24T12:28:19Z [18648] DEBUG database '_system' init/upgrade done
2017-12-24T12:28:19Z [18648] INFO database upgrade passed
[...]
```

So the `--database.auto-upgrade true` option seems to do its job, just the process ends then with no messages

As wrote above, if I then run again the process, without the option `-database.auto-upgrade` it stays up:

```
D:\>"D:\qa\3.3.0\ArangoDB3e-3.3.0-1_win64\usr\bin\arangod.exe" --database.directory  D:\qa\3.2.10\ArangoDB3e-3.2.10-1_win64\var\lib\arangodb3
2017-12-24T12:29:58Z [9252] INFO ArangoDB 3.3.0 enterprise [win64] 64bit, using VPack 0.1.30, RocksDB 5.6.0, ICU 58.1, V8 5.7.492.77, OpenSSL 1.0.2a 19 Mar 2015
2017-12-24T12:29:58Z [9252] INFO using storage engine mmfiles
2017-12-24T12:29:58Z [9252] INFO {cluster} Starting up with role SINGLE
2017-12-24T12:29:58Z [9252] INFO {authentication} Authentication is turned on (system only)
2017-12-24T12:29:58Z [9252] INFO using endpoint 'http+tcp://127.0.0.1:8529' for non-encrypted requests
2017-12-24T12:30:00Z [9252] INFO ArangoDB (version 3.3.0 enterprise [win64]) is ready for business. Have fun!
```

**Comment by sleto-it** [comment]
As far as I see, in `UpgradeFeature::start()` there is a call to force the shutdown `server()->beginShutdown();` :

https://github.com/arangodb/arangodb/blob/devel/arangod/RestServer/UpgradeFeature.cpp#L158

However, no messages are printed to inform that the server is being shutdown, and this might cause confusion as latest log entry is:

``` 
INFO ArangoDB (version 3.3.0 enterprise [win64]) is ready for business. Have fun!
````

We may consider to add a log entry, INFO, saying that the server is being shutdown and to restart it normally (without the `-database.auto-upgrade` option)

Documentation does not explicitly say that the server will shutdown. From 
https://docs.arangodb.com/3.3/Manual/Administration/Configuration/GeneralArangod.html :

> If the version number found in the database directory is lower than the version number the server is running, the server will check whether there are any upgrade tasks to perform. It will then execute all required upgrade tasks and print their statuses. If one of the upgrade tasks fails, the server will exit and refuse to start. Re-starting the server with the upgrade option will then again trigger the upgrade check and execution until the problem is fixed. If all tasks are finished, the server will start normally.

so if shutting down is expected, I believe we should consider citing that in the doc as well

**Comment by garyhai** [comment]
We deployed an arangodb server for development, which keep upgrade to the last version of arangodb. So the option of '--database.auto-upgrade' is sugar for maintenance of developing environment.

**Comment by goedderz** [comment]
Related: #391

**Comment by jsteemann** [comment]
Fixed by @goedderz in current devel.

**Comment by jsteemann** [comment]
Commit a7ee38b5fa78ff725f283817573f8b8200a1fccd

**Comment by maxkernbach** [comment]
Fixed in 3.3.4: When upgrading to 3.3.4 with `arangod --database.auto-upgrade true`
the following output message signals that the server is being shutdown.

```
arangod --database.auto-upgrade true
2018-03-14T15:49:49Z [17048] INFO detected operating system: Linux version 4.13.0-36-generic (buildd@lcy01-amd64-017) (gcc version 7.2.0 (Ubuntu 7.2.0-8ubuntu3.2)) #40-Ubuntu SMP Fri Feb 16 20:07:48 UTC 2018
2018-03-14T15:49:49Z [17048] INFO using storage engine mmfiles
2018-03-14T15:49:49Z [17048] INFO {authentication} Authentication is turned on (system only), authentication for unix sockets is turned on
2018-03-14T15:49:49Z [17048] INFO {syscall} file-descriptors (nofiles) hard limit is 8192, soft limit is 8192
2018-03-14T15:49:49Z [17048] INFO database upgrade passed
2018-03-14T15:49:49Z [17048] INFO {startup} Server will now shutdown due to upgrade, database init or admin restoration.

```
